lib/google/gax/operation.rb

```diff
@@ -0,0 +1,167 @@
+      end
+
+      # Checks if the operation is done. This does not send a new api call,
+      # but checks the result of the previous api call to see if done.
+      #
+      # @return [Boolean] Whether the operation is done.
+      def done?
+        @grpc_op.done
+      end
+
+      # Checks if the operation is done and the result is an error.
+      # If the operation is not finished then this will return false.
+      #
+      # @return [Boolean] Whether an error has been returned.
+      def error?
+        done? ? @grpc_op.result == :error : false
+      end
+
+      # Cancels the operation.
+      def cancel
+        @client.cancel_operation @grpc_op.name
+      end
+
+      # Reloads the operation object.
+      #
+      # @param backoff_settings [Google::Gax::BackoffSettings]
+      #   The backoff settings used to manipulate how this method retries
+      #   checking if the operation is done.
+      # @return [Google::Gax::Longrunning]
+      #   Since this method changes internal state, it returns itself.
+      def reload!(backoff_settings: nil)
+        options = CallOptions.new(
+          retry_options: RetryOptions.new(nil, backoff_settings)
+        )
+        @grpc_op = @client.get_operation @grpc_op.name, options: options
+        if done?
+          @callbacks.each { |proc| proc.call(self) }
+          @callbacks.clear
+        end
+        self
+      end
+      alias refresh! reload!
+
+      # Blocking method to wait until the operation has completed or the
+      # maximum timeout has been reached.
+      #
+      # @param backoff_settings [Google::Gax::BackoffSettings]
+      #   The backoff settings used to manipulate how this method retries
+      #   checking if the operation is done.
+      # @yield [Google::Rpc::Status | Object]
+      #   If a block is given, runs the block using the results
+      #   of the operation.
+      def wait_until_done!(backoff_settings: nil)
+        backoff_settings = BackoffSettings.new unless backoff_settings
+
+        delay = backoff_settings.initial_retry_delay_millis || 100
+        max_delay = backoff_settings.max_retry_delay_millis || 60_000
+        delay_multiplier = backoff_settings.retry_delay_multiplier || 1.3
+        total_timeout = backoff_settings.total_timeout_millis || 60_000
+        deadline = Time.now + total_timeout
+        until done?
+          sleep(rand(delay) / MILLIS_PER_SECOND)
+          if Time.now >= deadline
+            raise RetryError, 'Retry total timeout exceeded with exception'
+          end
+          delay = [delay * delay_multiplier, max_delay].min
+          reload!(backoff_settings: backoff_settings)
+        end
+        yield(results) if block_given?
+      end
+
+      # Registers a callback to be run when a refreshed operation is marked
+      # as done. If the operation has completed prior to a call to this function
+      # the callback will be called instead of registered.
+      def on_done(&block)
+        if done?
```

Comment:
I think this could address a theoretical situation where the initial return value of Operation could have `done: true` already.  I don't know this happens actually (probably quite unlikely) but if that happens, the `on_done` block won't run at all, which makes me a bit nervous.